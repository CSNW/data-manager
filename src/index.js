export { default as Store } from './store';
export { default as table } from './table';
export {
  filter,
  map,
  flatMap,
  sort,
  sortBy,
  clone,
  cloneSeries,
  groupBy,
  from
} from './operations';
export { compare, flow } from './utils';
