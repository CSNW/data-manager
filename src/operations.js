import {
  compare,
  mapValues,
  shallowCloneObj,
  flatMap as _flatMap
} from './utils';

/**
 * Filter series values to only selected rows
 *
 * @param {function} iterator (row, index, rows) => boolean
 * @returns {function} operation
 */
export function filter(iterator) {
  return data => mapValues(data, values => values.filter(iterator));
}

/**
 * Map series values
 *
 * @param {function} iterator (row, index, rows) => row
 * @returns {function} operation
 */
export function map(iterator) {
  return data => mapValues(data, values => values.map(iterator));
}

/**
 * Map series objects
 *
 * @param {function} iterator (series, index, data) => series
 * @returns {function} operation
 */
export function mapSeries(iterator) {
  return data => data.map(iterator);
}

/**
 * Flat map series values
 *
 * @param {function} iterator (row, index, rows) => row[]
 * @returns {function} operation
 */
export function flatMap(iterator) {
  return data =>
    data.map(series => {
      series.values = _flatMap(series.values, iterator);
      return series;
    });
}

/**
 * Sort series values
 *
 * @param {function} comparator (row_a, row_b) => number
 * @returns {function} operation
 */
export function sort(comparator) {
  return data => mapValues(data, values => values.sort(comparator));
}

/**
 * Sort series values by key
 *
 * @param {string} key
 * @param {function} comparator (a, b) => number
 * @returns {function} operation
 */
export function sortBy(key, comparator) {
  return sort((row_a, row_b) => comparator(row_a[key], row_b[key]));
}

/**
 * Clone series values
 *
 * @returns {function} operation
 */
export function clone() {
  return data => mapValues(data, values => values.map(shallowCloneObj));
}

/**
 * Clone series objects (does not clone underlying rows)
 *
 * @returns {function} operation
 */
export function cloneSeries() {
  return data => data.map(series => shallowCloneObj(series));
}

/**
 * Group series values by key
 *
 * @example
 * ```js
 * import { Store, table, groupBy } from 'data-manager';
 * import cast from 'data-manager/cast.macro';
 *
 * // data.csv
 * // a,b
 * // 1,a
 * // 2,b
 * // 3,b
 * // 4,a
 * const store = new Store();
 * const data = table('data.csv', cast({ a: Number, b: String }));
 *
 * async function grouped() {
 *   const results = await store.query(
 *     data,
 *     groupBy('b')
 *   );
 *
 *   // results = [
 *   //   { group: { b: 'a' }, values: [{ a: 1, b: 'a' }, { a: 4, b: 'a' }] },
 *   //   { group: { b: 'b' }, values: [{ a: 2, b: 'b' }, { a: 3, b: 'b' }] }
 *   // ]
 *
 *   const alternative = await store.query(
 *     data,
 *     groupBy('b', (b, series) => {
 *       // `values` is added directly to returned series
 *       series.key = b;
 *
 *       return series;
 *     })
 *   );
 *
 *   // alternative = [
 *   //   { key: 'a', values: [{ a: 1, b: 'a' }, { a: 4, b: 'a' }] },
 *   //   { key: 'b', values: [{ a: 2, b: 'b' }, { a: 3, b: 'b' }] }
 *   // ]
 * }
 * ```
 * @param {string} key
 * @param {function} [toSeries]
 */
export function groupBy(key, toSeries) {
  toSeries =
    toSeries ||
    ((value, series) => {
      series.group = series.group || {};
      series.group[key] = value;

      return series;
    });

  return data =>
    _flatMap(data, series => {
      const groups = new Map();
      series.values.forEach(row => {
        const id = row[key];

        if (!groups.has(id)) groups.set(id, []);
        const group = groups.get(id);

        group.push(row);
      });

      const grouped = [];
      groups.forEach((values, id) => {
        const group_series = toSeries(id, shallowCloneObj(series));
        group_series.values = values;

        grouped.push(group_series);
      });

      return grouped;
    });
}
