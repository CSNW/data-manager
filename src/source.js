export default class Source {
  fetch(path, map) {
    // TODO
  }

  query() {
    // TODO
  }
}
