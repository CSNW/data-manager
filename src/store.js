import { csv as fetchCsv } from 'd3-fetch';

export default class Store {
  constructor() {
    this.cache = new Map();
    this.loading = new Map();
  }

  load(path, convert) {
    if (this.cache.has(path)) {
      return Promise.resolve(this.cache.get(path));
    }
    if (this.loading.has(path)) {
      return this.loading.get(path);
    }

    const loading = Promise.resolve(this.fetch(path))
      .then(raw => {
        return convert ? raw.map(convert).filter(Boolean) : raw;
      })
      .then(values => {
        this.cache.set(path, values);
        this.loading.delete(path);

        return values;
      });

    this.loading.set(path, loading);

    return loading;
  }

  fetch(path) {
    return fetchCsv(path);
  }

  query(...operations) {
    return operations.reduce((loading, operation) => {
      return loading.then(data => operation(data, this));
    }, Promise.resolve([]));
  }
}
