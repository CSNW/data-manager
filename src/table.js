export default function table(path, map) {
  return (_, source) => {
    return source.fetch(path, map).then(values => ({ values }));
  };
}
