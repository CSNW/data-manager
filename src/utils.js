export function mapValues(data, iterator) {
  return data.map(series => {
    const mapped = shallowCloneObj(series);
    mapped.values = iterator(series.values);

    return mapped;
  });
}

export function shallowCloneObj(obj) {
  const cloned = {};
  for (const key in obj) {
    cloned[key] = obj[key];
  }
  return cloned;
}
