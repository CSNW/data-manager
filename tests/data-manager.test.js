import { Store, table, filter, map, sortBy, compare } from '../src/index';
import cast, { derived } from '../cast.macro';

import { join } from 'path';
import { promisify } from 'util';
const readFile = promisify(require('fs').readFile);
import { csvParse } from 'd3-dsv';

class LocalStore extends Store {
  async fetch(path, convert) {
    path = join(__dirname, '__fixtures__', path);

    const data = await readFile(path);
    const values = csvParse(data.toString(), convert);

    return values;
  }
}

const single = table(
  'data/single.csv',
  cast({
    a: Number,
    b: Number
  })
);

test('should fetch and filter csv data', async () => {
  const store = new LocalStore();

  const results = await store.query(single, filter(row => row.a > 2));

  expect(results).toMatchSnapshot();
});

test('should fetch, filter, map, and sort csv data', async () => {
  const store = new LocalStore();

  const results = await store.query(
    single,
    filter(row => row.a <= 4),
    map(row => {
      row.computed = row.a + row.b;
      return row;
    }),
    sortBy('a', compare.numbersDescending)
  );

  expect(results).toMatchSnapshot();
});

test('should cast csv data with macro', async () => {
  const store = new LocalStore();

  const results = await store.query(
    table(
      'data/single.csv',
      cast({
        a: value => `a = ${value}`,
        b: value => `b = ${value}`,
        c: derived(row => Number(row.a) + Number(row.b))
      })
    )
  );

  expect(results).toMatchSnapshot();
});
