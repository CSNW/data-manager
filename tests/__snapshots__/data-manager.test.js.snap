// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should cast csv data with macro 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": "a = 0",
        "b": "b = 10",
        "c": 10,
      },
      Object {
        "a": "a = 1",
        "b": "b = 8",
        "c": 9,
      },
      Object {
        "a": "a = 2",
        "b": "b = 6",
        "c": 8,
      },
      Object {
        "a": "a = 3",
        "b": "b = 4",
        "c": 7,
      },
      Object {
        "a": "a = 4",
        "b": "b = 2",
        "c": 6,
      },
      Object {
        "a": "a = 5",
        "b": "b = 0",
        "c": 5,
      },
    ],
  },
]
`;

exports[`should fetch and filter csv data 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 5,
        "b": 0,
      },
    ],
  },
]
`;

exports[`should fetch, filter, map, and sort csv data 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 4,
        "b": 2,
        "computed": 6,
      },
      Object {
        "a": 3,
        "b": 4,
        "computed": 7,
      },
      Object {
        "a": 2,
        "b": 6,
        "computed": 8,
      },
      Object {
        "a": 1,
        "b": 8,
        "computed": 9,
      },
      Object {
        "a": 0,
        "b": 10,
        "computed": 10,
      },
    ],
  },
]
`;

exports[`should match csv data with macro 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 2,
        "b": 6,
      },
    ],
  },
]
`;
